hydromt/io/path.py

```diff
@@ -0,0 +1,147 @@
+"""A module to handle paths from different platforms in a cross platform compatible manner."""
+from os import PathLike, getcwd, getlogin
+from os.path import abspath, exists, join
+from pathlib import Path, PurePosixPath, PureWindowsPath
+from platform import system
+from typing import List, Optional, Tuple
+
+
+def _posix_is_mounted(path) -> bool:
+    # check if the first dir is one of the canonical
+    # places to mount in linux
+    return Path(path).parts[1] in ["mnt", "media"]
+
+
+def _count_slashes(path) -> Tuple[int, int]:
+    """Return the number of each shales in the path as (forward,backward)."""
+    str_path = str(path)
+    forward_count = str_path.count("/")
+    backward_count = str_path.count("\\")
+
+    return forward_count, backward_count
+
+
+def _describes_windows_path(path: PathLike) -> bool:
+    forward_count, backward_count = _count_slashes(path)
+    # not great, but it's the best we have
+    return forward_count < backward_count
+
+
+def _describes_posix_path(path: PathLike) -> bool:
+    forward_count, backward_count = _count_slashes(path)
+    # not great, but it's the best we have
+    return forward_count > backward_count
+
+
+def make_path_abs_and_cross_platform(path: PathLike) -> Path:
+    r"""
+    Return an absolute path accodring to the conventions of the platform we're running on.
+
+    Dealing with paths that describe a file path for a platrom we're not on (e.g. handeling Windows paths on a POSIX system)
+    easily leads to trouble. This function will apply the following assumptions:
+
+    - Paths that describe paths on the same platform are handled like normally
+    - on POSIX platforms a windows path like C:\\to\\somehwere maps to /mnt/c/to/somewhere. for linux this is the cannonical place to mount things like network shares.
+    - Relative Windows paths on POSIX platforms are simply converted to their posix/linux equivalent and used like normal
+    - on Windows a POSIX path like /mnt/d/to/somewhere will map to D:\\to\\somewhere mostly as an inverse as the previous point.
+    - A path like /home/user/to/somewhere will map to C:\\Users\\{login_name_of_user}\\to\\somewhere. so keep in mind it will map to the
+        home dir of the current user, not the user specified in the path.
```

Comment:
Do I understand correctly that this function allows a user on linux to provide a windows-like path which is transformed to a linux path based on conventions of how folders on these platforms art typically organized? 

To be frank, I don't yet see the case for this in relation to a model root. A user should, in my opinion, simply supply a path that works on the local platform they are working on. Can you motivate the case a bit more?

For a data catalog this is different. We now have a single root per data catalog file and hence the same catalog file will not work on multiple OS platforms. But rather than automatically transforming the path I would be in favor of implementing a mechanism to support either overwriting the root that is specified in the file or setting a mapping/list of multiple roots in the file. This would provide more flexibility to where data is stored on different systems. For instance the infamous hydromt p-drive on windows is "p:\wflow_global\hydromt\", if accessed from our hpc clusters "/p/wflow_global/hydromt" and if accessed from wsl we would have to mount it manually and hence the path also depends on how the users organizes their environment (typically this would be "/mnt/p/wflow_global/hydromt" as assumed here).

Reply:
> To be frank, I don't yet see the case for this in relation to a model root. A user should, in my opinion, simply supply a path that works on the local platform they are working on. Can you motivate the usecase a bit more? E.g. when would a user supply a POSIX path to access a model when working on Windows?

I am the case. We have a data catalog (the deltares data catalog) that provides all windows paths, but I work on linux. If I change the path in the deltares data catalog to a POSIX path then all the windows users will be unhappy. Having to change the path but not commit it every time I want to use it is too much wasted effort. Therefore cross platform teams are the motivating case here. This way you just have to provide a common mount point and the catalog will continue to work. Perhaps everyone at Deltares works on windows but that's not an assumption we should make, since things like cloud environments or docker containers almost exclusively run on POSIX systems and we don't have to have to make different catalogs for the cloud if we even watn to move in that direction. So we need _a_ mechanism for accessing paths on multiple platforms. 

I'm fine with adding the possibility of providing multiple roots or a way to override it, but I have had to spend much to long debugging weird cross platform path bugs so we have to provide SOME sollution for this, even if it's just for my sanity.

Reply:
So we agree there is an issue with the deltares DataCatalog root for Linux users. Can you also clarify the issue the model root? Just trying to understand it in order to be able to review.

For the DataCatalog I would be in favor of dealing with this in a more explicit and flexible manner so it does not only support wsl but also the deltares hpc (see my earlier comment) and other potential use cases. Let's make a separate issue for this.

Reply:
I'm not actually sure what you're asking, but as you mentioned making a seperate issue for this, I think it would be best if we discuss this face to face and then indeed make a seperate issue. I want to continue iwth the model region as well, so I'll be merging this PR, but we can discuss any changes you like to see next week.